### PR TITLE
Remove AnalysisService and reduce it to a simple name to analyzer mapping

### LINKED ISSUE #19828: Simplify AnalysisService
We want to simplify `AnlaysisService` so we keep fewer objects on heap per index. This means removing the `tokenizer`, `charFilter`, and `tokenFilter` methods and their support from the `AnalysisService`. Those methods are only used to build analyzers, both during analysis service startup and during the `_analyze` API. I've opened #19827 to stop using those methods in the `_analyze` API. After that is done, we can work on this to remove the methods entirely.

### PR DESCRIPTION
**Pull request by s1monw**
Today we hold on to all possible tokenizers, tokenfilters etc. when we create
an index service on a node. This was mainly done to allow the `_analyze` API to
directly access all these primitive. We fixed this in #19827 and can now get rid of
the AnalysisService entirely and replace it with a simple map like class. This
ensures we don't create a gazillion long living objects that are entirely useless since
they are never used in most of the indices. Also those objects might consume a considerable
amount of memory since they might load stopwords or synonyms etc.

Closes #19828

### CODE REVIEW

**Comment by s1monw** [comment]
@nik9000 can you take a look - also @imotov you might be interested in this

**Comment by nik9000** [comment]
I've added this to my review queue. It is big enough I'll wait until I have some coffee in me before I try and fit it in my head.

**Comment by s1monw** [comment]
retest this please

File: core/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @Override
+    public void close() {
+        for (NamedAnalyzer analyzer : analyzers.values()) {
+            if (analyzer.scope() == AnalyzerScope.INDEX) {
+                try {
+                    analyzer.close();
+                } catch (NullPointerException e) {
+                    // because analyzers are aliased, they might be closed several times
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java]: Isn't `close()` supposed to be idempotent?

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java [hunk 2]
```diff
@@ -112,9 +116,25 @@ private void checkMappingsCompatibility(IndexMetaData indexMetaData) {
             // been started yet. However, we don't really need real analyzers at this stage - so we can fake it
             IndexSettings indexSettings = new IndexSettings(indexMetaData, this.settings);
             SimilarityService similarityService = new SimilarityService(indexSettings, Collections.emptyMap());
+            final NamedAnalyzer fakeDefault = new NamedAnalyzer("fake_default", new Analyzer() {
+                @Override
+                protected TokenStreamComponents createComponents(String fieldName) {
+                    throw new UnsupportedOperationException("shouldn't be here");
+                }
+            });
+            final Map<String, NamedAnalyzer> analyzerMap =    new AbstractMap<String, NamedAnalyzer>() {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java]: Extra ` `s

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java [hunk 3]
```diff
@@ -112,9 +116,25 @@ private void checkMappingsCompatibility(IndexMetaData indexMetaData) {
...
+                @Override
+                public NamedAnalyzer get(Object key) {
+                    return new NamedAnalyzer((String)key, fakeDefault.analyzer());
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java]: `toString` on the key?

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java]: It is probably worth adding a comment about why we need to support doing this for *any* string. This map's `entrySet` method doesn't agree with the `get` which is probably fine for this but we should explain it.

File: core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java [hunk 4]
```diff
@@ -399,4 +421,132 @@ public void close() throws IOException {
             IOUtils.close(analyzerProviderFactories.values().stream().map((a) -> ((PreBuiltAnalyzerProviderFactory)a).analyzer()).collect(Collectors.toList()));
         }
     }
+
+    public IndexAnalyzers build(IndexSettings indexSettings,
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: Note for anyone else reviewing: these next two methods look like they are just copied from the (now removed) `AnalysisService`.

File: core/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java]: If we do need to keep this clause can you add an assertion that the index's version created setting is less than 5.0?

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java (hunk 3 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java]: NOOOO I want to fail if it's not a String :)

File: core/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java]: isn't this a preexiting issue bikeshedder :)

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java (hunk 3 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java]: ok ok ok

File: core/src/main/java/org/elasticsearch/index/mapper/ParseContext.java [hunk 5]
```diff
@@ -29,15 +29,11 @@
...
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.xcontent.XContentParser;
-import org.elasticsearch.index.analysis.AnalysisService;
 
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 
-/**
- *
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/mapper/ParseContext.java]: 👍

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 6]
```diff
@@ -868,4 +874,22 @@ public static IndicesModule newTestIndicesModule(Map<String, Mapper.TypeParser>
             }
         ));
     }
+
+    public static final class TestAnalysis {
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java]: Cute! Probably worth adding javadoc for it.

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 7]
```diff
@@ -812,33 +816,35 @@ protected static long spinForAtLeastOneMillisecond() {
...
+                                                  AnalysisPlugin... analysisPlugins) throws IOException {
         Settings indexSettings = Settings.builder().put(settings)
                 .put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT)
                 .build();
-        return createAnalysisService(IndexSettingsModule.newIndexSettings(index, indexSettings), nodeSettings, analysisPlugins);
+        return createTestAnalysis(IndexSettingsModule.newIndexSettings(index, indexSettings), nodeSettings, analysisPlugins);
     }
 
     /**
      * Creates an AnalysisService with all the default analyzers configured.
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java]: Out of date doc.

**Comment by nik9000** [comment]
Despite the number of files this touches it is fairly strait forward. I think it makes the code easier to understand. I left a bunch of minor stuff.

File: core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java (hunk 4 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: they are

**Comment by s1monw** [comment]
@nik9000 @rjernst I updated the PR

**Review by nik9000 — APPROVED** [review]

**Comment by s1monw** [comment]
ported to 5.x in 2d70d46

**Pull request merged**